tests/abstraction_ak_tests.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::convert::{TryFrom, TryInto};
+
+use tss_esapi::{
+    abstraction::{ak, ek},
+    constants::{
+        algorithm::{AsymmetricAlgorithm, Cipher, HashingAlgorithm, SignatureScheme},
+        types::session::SessionType,
+    },
+    handles::AuthHandle,
+    structures::{Auth, Digest},
+    utils::TpmaSessionBuilder,
+};
+
+mod common;
+use common::create_ctx_without_session;
+
+#[test]
+fn test_create_ak_rsa_rsa() {
+    let mut context = create_ctx_without_session();
+
+    let ek_rsa = ek::create_ek_object(&mut context, AsymmetricAlgorithm::Rsa).unwrap();
+    ak::create_ak(
+        &mut context,
+        ek_rsa,
+        AsymmetricAlgorithm::Rsa,
+        HashingAlgorithm::Sha256,
+        SignatureScheme::RsaPss,
+        None,
+    )
+    .unwrap();
+}
+
+#[test]
+fn test_create_ak_rsa_ecc_invalid_sig_alg() {
+    let mut context = create_ctx_without_session();
+
+    let ek_rsa = ek::create_ek_object(&mut context, AsymmetricAlgorithm::Rsa).unwrap();
+    if ak::create_ak(
+        &mut context,
+        ek_rsa,
+        AsymmetricAlgorithm::Ecc,
+        HashingAlgorithm::Sha256,
+        SignatureScheme::RsaPss,
+        None,
+    )
+    .is_ok()
+    {
+        // We can't use unwrap_err because that requires Debug on the T
+        panic!("Should have errored");
+    }
+}
+
+#[test]
+fn test_create_ak_rsa_ecc() {
+    let mut context = create_ctx_without_session();
+
+    let ek_rsa = ek::create_ek_object(&mut context, AsymmetricAlgorithm::Rsa).unwrap();
+    if ak::create_ak(
+        &mut context,
+        ek_rsa,
+        AsymmetricAlgorithm::Ecc,
+        HashingAlgorithm::Sha256,
+        SignatureScheme::EcDsa,
+        None,
+    )
+    .is_ok()
+    {
+        // We can't use unwrap_err because that requires Debug on the T
+        panic!("Should have errored");
+    }
+}
+
+#[test]
+fn test_create_and_use_ak() {
```

Comment:
This test is a full end-to-end test of the common way that an EK and AK are used: encrypt the secret against the `EK`, but add the name of the `AK` to ensure the TPM verifies that the `AK` object is known under the expected `Name` (i.e. with the correct attributes, since those are part of its `Name`).
This happens to all be on the same TPM, but normally the `MakeCredential` part would happen on the server side based on the `EK` and `AK` received from a TPM.

This is intended to implement the `Credential Protection` protocol (chapter 24 of `Trusted Platform Module Library Part 1: Architecture Level 00 Revision 01.59`).

Reply:
This is a technique a server can use to ensure that the particular `AK` that it got is created on a TPM that also has an `EK` it can store, with attributes it wants (like for example `FixedTPM` and `sensitiveDataOrigin` both set to ensure it's non-exportable and created by the TPM). The server can send a `MakeCredential`'d challenge to the `Name`, and then the TPM will only decrypt it if it has the expected object, so if the client can send the challenge back, that means it must be a good object. Then the server can from then on use the `AK` for requesting signed statements (i.e. quotes), because the `EK` is decrypt-only, and can't be used to sign.  (which is done to incentivise privacy - after this protocol, the server can't know for sure that the TPM it's talking to is the TPM with the earlier `EK`, but it knows it's talking to the same TPM that generated the `AK`).